Docs: Refactor the build & build customization pages (Diátaxis) (#10028)
readthedocs/readthedocs.org: docs/user/build-customization.rst

Build process customization
===========================

Read the Docs has a :doc:`well-defined build process </builds>` that works for many projects.
We also allow customization of builds in primary ways:

`Extend the build process`_

>     This is useful if you want to use our deafult build process,
>     but just run additional commands.

Reviewer: ```suggestion
    Keep using the default build process,
    adding your own additional commands.
```

`Override the build process`_
    This option gives you *full control* over your build.
    Read the Docs supports any tool that generates HTML.

Extend the build process
------------------------

In the normal build process,

> the pre-defined jobs ``checkout``, ``system_dependencies``, ``create_environment``, ``install``, ``build`` and ``upload`` are executed.

Reviewer: Wouldn't this ideally be "steps" and then the steps consist of a list of "jobs" or "commands"?

Reviewer 2: We talked about this already. I don't have the PR at hand, but there were some conversation around this. These are called "jobs" and the "steps" are each individual commands.

Reviewer: Noted!

Reviewer: It seems from @humitos explanation that we can align with how other CIs understand jobs, steps, pipelines etc... This is a good source: https://www.jenkins.io/doc/book/pipeline/#step

> Read the Docs also exposes additional ujobs,
> which allow you to customize the build process by running shell commands.

Reviewer: ```suggestion
Read the Docs also exposes these jobs,
which allows you to customize the build process by adding shell commands.
```

Author: Bleh, not sure why my editor didn't flag these typos 🙃

The jobs where users can customize our default build process are:

.. list-table::
   :header-rows: 1

>    :widths: 25, 75

Reviewer: ```suggestion
   :widths: 25 75
```

``build.jobs`` examples
~~~~~~~~~~~~~~~~~~~~~~~

We've included some common examples where using :ref:`config-file/v2:build.jobs` will be useful.
These examples may require some adaptation for each projects' use case,
we recommend you use them as a starting point.

Unshallow git clone
^^^^^^^^^^^^^^^^^^^

Read the Docs does not perform a full clone on ``checkout`` job to reduce network data and speed up the build process.
Because of this, extensions that depend on the full Git history will fail.

> To avoid this, it's possible to unshallow the git clone:

Reviewer: ```suggestion
To avoid this, it's possible to unshallow the :program:`git clone`:
```

   `the Unix implementation does this automatically <https://tldp.org/LDP/abs/html/exitcodes.html>`_
   for exit codes greater than 255.

>    ::

Reviewer: Didn't this work with `pycon`?

Author: It was giving me a `black` error, but I thought `pycon` was a typo :D

Reviewer: Yes, I thought so too a long time ago... but it's a real thing :) https://pygments.org/docs/lexers/#pygments.lexers.python.PythonConsoleLexer

Author: ```suggestion
   .. code-block:: pycon
```

.. warning::

   This feature is in *beta* and could change without warning.
   It does not yet support some of the Read the Docs' features like the :term:`flyout menu`.
   We do our best to not break existing configurations,
   but use this feature at your own risk.

If your project requires full control of the build process,
and :ref:`extending the build process <build-customization:extend the build process>` is not enough,
all the commands executed during builds can be overridden using the :ref:`config-file/v2:build.commands`.

As Read the Docs does not have control over the build process,
you are responsible for running all the commands required to install requirements and build your project.

Where to put files
~~~~~~~~~~~~~~~~~~

It is your responsibility to generate HTML and other formats of your documentation using :ref:`config-file/v2:build.commands`.

> The contents of the ``_readthedocs/<format>/`` directory will be hosted as part of your documentation.

Reviewer: ```suggestion
The contents of the ``$READTHEDOCS_OUTPUT/<format>/`` directory will be hosted as part of your documentation. ``$READTHEDOCS_OUTPUT`` is an :doc:`environment variable </reference/environment-variables>` that references the location where outputs should be stored.
```

Reviewer: @humitos did we document `$READTHEDOCS_OUTPUT` anywhere yet?

Reviewer 2: Nope, we opened this issue to track this work: https://github.com/readthedocs/readthedocs.org/issues/9894

Author: I'd want to have it documented at least in the Env Vars section, before we put it here.

Reviewer: Yes, this little change can stand alone.

No one is assigned to #9894, but I don't mind doing it :+1:

Reviewer 2: @ericholscher sounds good to me 👍🏼

> Supported :ref:`formats <downloadable-documentation:accessing offline formats>` are:
> 
> * ``html``
> * ``htmlzip``
> * ``pdf``
> * ``epub``

Reviewer: ```suggestion
Supported :ref:`formats <downloadable-documentation:accessing offline formats>` are ``html``, ``htmlzip``, ``pdf``, and ``epub`` and following output targets are published if they exist:

* ``_readthedocs/html/`` (required)
* ``_readthedocs/htmlzip/``
* ``_readthedocs/pdf/``
* ``_readthedocs/epub/``
```

Reviewer: @humitos it's correct that HTML is required?

> * ``_readthedocs/epub/``

Reviewer 2: This can be updated now to `$READTHEDOCS_OUTPUT/<format>` 🎉

Reviewer: :sunglasses:

Author: I think we probably want to keep it explicit, and mention the variable as an option?

Reviewer 2: My point is that this may change in the future and people shouldn't use these path at all. They should use the variables. Otherwise, it could break in the future.

Reviewer: I guess you also mentioned new environment variables @humitos - something like `$READTHEDOCS_OUTPUT_DIR_HTML` and `$READTHEDOCS_OUTPUT_DIR_PDF` that would be nicer for people to use.

So let's wait until that work is done :+1:
